Mathlib/RingTheory/PowerSeries/Substitution.lean

```diff
@@ -0,0 +1,293 @@
+  IsNilpotent (MvPowerSeries.constantCoeff τ S a)
+
+theorem hasSubst_iff {a : MvPowerSeries τ S} :
+  HasSubst a ↔ MvPowerSeries.HasSubst (Function.const Unit a) :=
+  ⟨fun ha ↦ MvPowerSeries.hasSubst_of_constantCoeff_nilpotent (Function.const Unit ha),
+   fun ha  ↦ (ha.const_coeff ())⟩
+
+theorem HasSubst.const {a : MvPowerSeries τ S} (ha : HasSubst a) :
+    MvPowerSeries.HasSubst (fun () ↦ a) :=
+  hasSubst_iff.mp ha
+
+theorem hasSubst_iff_hasEval_of_discreteTopology
+    [TopologicalSpace S] [DiscreteTopology S] {a : MvPowerSeries τ S} :
+    HasSubst a ↔ PowerSeries.HasEval a := by
+  rw [hasSubst_iff, MvPowerSeries.hasSubst_iff_hasEval_of_discreteTopology, hasEval_iff]
+  rfl
+
+theorem HasSubst.hasEval [TopologicalSpace S] {a : MvPowerSeries τ S} (ha : HasSubst a) :
+    HasEval a := by
+  rw [hasEval_iff]
+  apply MvPowerSeries.HasSubst.hasEval
+  simpa [hasSubst_iff] using ha
+
+theorem HasSubst.of_constantCoeff_zero {a : MvPowerSeries τ S}
+    (ha : MvPowerSeries.constantCoeff τ S a = 0) : HasSubst a := by
+  simp [HasSubst, ha]
+
+/-- A variant of `HasSubst.of_constantCoeff_zero` for `PowerSeries`
+to avoid the expansion of `Unit` -/
+theorem HasSubst.of_constantCoeff_zero' {a : PowerSeries S}
+    (ha : PowerSeries.constantCoeff S a = 0) : HasSubst a :=
+  HasSubst.of_constantCoeff_zero ha
+
+protected theorem HasSubst.X (t : τ) :
+    HasSubst (MvPowerSeries.X t : MvPowerSeries τ S) := by
+  simp [HasSubst]
+
+/-- The univariate `X : R⟦X⟧` can be substituted in power series
+
+This lemma is added because `simp` doesn't find it from `HasSubst.X` -/
+protected theorem HasSubst.X' : HasSubst (X : R⟦X⟧) :=
+  HasSubst.X _
+
+protected theorem HasSubst.X_pow {n : ℕ} (hn : n ≠ 0) : HasSubst (X ^ n : R⟦X⟧) :=
+  HasSubst.of_constantCoeff_zero' (by simp [hn])
+
+protected theorem HasSubst.monomial {n : τ →₀ ℕ} (hn : n ≠ 0) (s : S) :
+    HasSubst (MvPowerSeries.monomial S n s) := by
+  classical
+  apply HasSubst.of_constantCoeff_zero
+  rw [← MvPowerSeries.coeff_zero_eq_constantCoeff, MvPowerSeries.coeff_monomial,
+    if_neg hn.symm]
+
+/-- A variant of `HasSubst.monomial` to avoid the expansion of `Unit` -/
+protected theorem HasSubst.monomial' {n : ℕ} (hn : n ≠ 0) (s : S) :
+    HasSubst (monomial S n s) :=
+  HasSubst.monomial (Finsupp.single_ne_zero.mpr hn) s
+
+theorem HasSubst.zero : HasSubst (0 : MvPowerSeries τ R) := by
+  rw [hasSubst_iff]
+  exact MvPowerSeries.HasSubst.zero
+
+/-- A variant of `HasSubst.zero` to avoid the expansion of `Unit` -/
+theorem HasSubst.zero' : HasSubst (0 : PowerSeries R) :=
+  PowerSeries.HasSubst.zero
+
+theorem HasSubst.add {f g : MvPowerSeries τ R} (hf : HasSubst f) (hg : HasSubst g) :
+    HasSubst (f + g) := by
+  simp only [HasSubst, map_add]
+  exact Commute.isNilpotent_add (Commute.all _ _) hf hg
+
+theorem HasSubst.mul_left {f g : MvPowerSeries τ R} (hf : HasSubst f) :
+    HasSubst (f * g) := by
+  simp only [HasSubst, map_mul]
+  exact Commute.isNilpotent_mul_left (Commute.all _ _) hf
```

Comment:
```suggestion
theorem HasSubst.add {f g : MvPowerSeries τ R} (hf : HasSubst f) (hg : HasSubst g) :
    HasSubst (f + g) :=
  (Commute.all _ _).isNilpotent_add hf hg
```
You can probably do the same in various places. Also, you repeat a lot of times `f g : MvPowerSeries τ R` (sometimes with different names), can you use a `variable` line?

Reply:
I don't understand how it works for `add`, but not for `mul`!